File: algorithm/프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.Collections;
import java.util.PriorityQueue;

class Solution {
    public long solution(int n, int[] works) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for(int work: works) {
            pq.add(work);
        }
        
        for (int i = 0; i < n; i++) {
            int work = pq.poll();
            if (work == 0) {
                return 0;
            }
            
            work -= 1;
            pq.add(work);
        }
                
        long answer = 0;
        for (int remain: pq) {
            answer += Math.pow(remain, 2);
        }
        
        return answer;
    }
}

File: algorithm/프로그래머스/3/12927. 야근 지수/야근 지수.py
import heapq

def solution(n, works):
    answer = 0
    
    hq = [-1 * work for work in works]
    heapq.heapify(hq)
    
    for _ in range(n):
        work = heapq.heappop(hq)
        if work == 0:
            return 0
        
        work += 1
        heapq.heappush(hq, work)
        
    for remain_work in hq:
        answer += (-1 * remain_work) ** 2
    
    return answer